homeassistant/components/lifx/light.py

```diff
@@ -608,7 +613,8 @@ async def update_hass(self, now=None):
         """Request new status and push it to hass."""
         self.postponed_update = None
         await self.async_update()
-        self.async_write_ha_state()
+        if self.entity_id is not None:
```

Comment:
Honestly I have no idea, but it does. I have some theories around possible race conditions that could cause this, but given how much more reliable reconnection is, I don't have to spend the hours trying to work it out. It just self heals already.

Reply:
`self.entity_id` should only ever be `None` if this function is called before the entity is added to hass

Reply:
Well, it is. Logs incoming.

Reply:
I'm going to resolve this conversation as the logs show that adding this check bypasses the error that occurs without it, but I will keep investigating to try and determine why it happens in the first place.

Reply:
This is a bug. Its registering the bulb multiple times because control is returned to the event loop in the else block of `register_bulb` which allows another instance of it to be called before the entity is in `self.entities`:

The check here is stale 
```python
    async def register_bulb(self, bulb):
        """Handle LIFX bulb registration lifecycle."""
        if bulb.mac_addr in self.entities:
```

when the bulb gets discovered again before the code in else block completes

```python3
        else:
            _LOGGER.debug("Connecting to %s (%s)", bulb.ip_addr, bulb.mac_addr)
....
```

Reply:
Which is why I've added a check before we even get here.

Reply:
The above is the cause of the LIFX does not generate unique id errors in your first log.

For this specific issue, the race is the other way around:

The `async_add_entities` creates a task under the hood so if `register_bulb` gets called again, the entity isn't finished adding yet which means it calls the update function prematurely.

Reply:
Both of these are resolved by limiting inflight discovery and it has the additional benefit of reducing network traffic.

Reply:
I want to refactor the entire integration and move the LIFX Manager and discovery out of the light entity so I can support config and diagnostic entities as well, but this at least fixes the current bugs.

Reply:
> Both of these are resolved by limiting inflight discovery and it has the additional benefit of reducing network traffic.

It doesn't seem like it's resolved if we still need this check

Reply:
You're right, this check isn't required any more. I think I fixed this first, then never came back to it after adding the inflight check, which resolved it as well.